src/meta/expr.lean

```diff
@@ -830,6 +821,67 @@ private meta def all_implicitly_included_variables_aux
 meta def all_implicitly_included_variables (es vs : list expr) : list expr :=
 all_implicitly_included_variables_aux es vs [] ff
 
+/-- Infer the type of an application of the form `f x1 x2 ... xn`, where `f` is an identifier.
+This also works if `x1, ... xn` contain free variables. -/
+protected meta def simple_infer_type (env : environment) (e : expr) : exceptional expr := do
+(@const tt n ls, es) ← return e.get_app_fn_args |
+  exceptional.fail "expression is not a constant applied to arguments",
+d ← env.get n,
+return $ (d.type.instantiate_pis es).instantiate_univ_params $ d.univ_params.zip ls
+
+/-- Auxilliary function for `head_eta_expand`. -/
+meta def head_eta_expand_aux : ℕ → expr → expr → expr
+| (n+1) e (pi x bi d b) :=
+lam x bi d $ head_eta_expand_aux n e b
+| _ e _ := e
+
+/-- `head_eta_expand n e t` eta-expands `e` `n` times, with the binders info and domains obtained
+  by its type `t`. -/
+meta def head_eta_expand (n : ℕ) (e t : expr) : expr :=
+((e.lift_vars 0 n).mk_app $ (list.range n).reverse.map var).head_eta_expand_aux n t
+
+/-- `e.eta_expand env dict` eta-expands all expressions that have as head a constant `n` in
+`dict`. They are expanded until they are applied to one more argument than the maximum in
+`dict.find n`. -/
+protected meta def eta_expand (env : environment) (dict : name_map $ list ℕ) : expr → expr
+| e := e.replace $ λ e _,
+  let (e0, es) := e.get_app_fn_args in
+  let ns := (dict.find e0.const_name).iget in
+  if ns = [] then none else
+  let e' := e0.mk_app $ es.map eta_expand in
+  let needed_n := ns.foldr max 0 + 1 in
+  if needed_n ≤ es.length then
+  some e' else do
+  e'_type ← (e'.simple_infer_type env).to_option,
+  some $ head_eta_expand (needed_n - es.length) e' e'_type
```

Comment:
It was a bit weird to read this code because it nests very deep to the right. A `do` block hides the nesting somewhat:
```suggestion
| e := e.replace $ λ e _, do
  let (e0, es) := e.get_app_fn_args,
  let ns := (dict.find e0.const_name).iget,
  guard (!ns.empty),
  let e' := e0.mk_app $ es.map eta_expand,
  let needed_n := ns.foldr max 0 + 1,
  if needed_n ≤ es.length then some e'
  else do
    e'_type ← (e'.simple_infer_type env).to_option,
    some $ head_eta_expand (needed_n - es.length) e' e'_type
```